src/Storages/MergeTree/CommonANNIndexes.cpp

```diff
@@ -0,0 +1,508 @@
+}
+
+float ANNCondition::getComparisonDistance() const
+{
+    if (where_query_type)
+    {
+        return ann_expr->distance;
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Not supported method for this query type");
+}
+
+std::vector<float> ANNCondition::getTargetVector() const
+{
+    return ann_expr->target;
+}
+
+String ANNCondition::getColumnName() const
+{
+    return ann_expr->column_name;
+}
+
+String ANNCondition::getMetric() const
+{
+    return ann_expr->metric_name;
+}
+
+size_t ANNCondition::getSpaceDim() const
+{
+    return ann_expr->target.size();
+}
+
+float ANNCondition::getPForLpDistance() const
+{
+    return ann_expr->p_for_lp_dist;
+}
+
+bool ANNCondition::queryHasWhereClause() const
+{
+    return where_query_type;
+}
+
+bool ANNCondition::queryHasOrderByClause() const
+{
+    return order_by_query_type;
+}
+
+UInt64 ANNCondition::getLimitCount() const
+{
+    if (index_is_useful)
+    {
+        return limit_count;
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "No LIMIT section in query, not supported");
+}
+
+String ANNCondition::getSettingsStr() const
+{
+    return ann_index_params;
+}
+
+void ANNCondition::buildRPN(const SelectQueryInfo & query)
+{
+    const auto & select = query.query->as<ASTSelectQuery &>();
+
+    if (select.prewhere()) // If query has PREWHERE section
+    {
+        traverseAST(select.prewhere(), rpn_prewhere_clause);
+    }
+
+    if (select.where()) // If query has WHERE section
+    {
+        traverseAST(select.where(), rpn_where_clause);
+    }
+
+    if (select.limitLength()) // If query has LIMIT section
+    {
+        traverseAST(select.limitLength(), rpn_limit_clause);
+    }
+
+    if (select.orderBy()) // If query has ORDERBY section
+    {
+        traverseOrderByAST(select.orderBy(), rpn_order_by_clause);
+    }
+
+    // Reverse RPNs for conveniences during parsing
+    std::reverse(rpn_prewhere_clause.begin(), rpn_prewhere_clause.end());
+    std::reverse(rpn_where_clause.begin(), rpn_where_clause.end());
+    std::reverse(rpn_order_by_clause.begin(), rpn_order_by_clause.end());
+}
+
+void ANNCondition::traverseAST(const ASTPtr & node, RPN & rpn)
+{
+    // If the node is ASTFUunction, it may have children nodes
+    if (const auto * func = node->as<ASTFunction>())
+    {
+        const ASTs & args = func->arguments->children;
+        // Traverse children nodes
+        for (const auto& arg : args)
+        {
+            traverseAST(arg, rpn);
+        }
+    }
+
+    RPNElement element;
+    // Get the data behind node
+    if (!traverseAtomAST(node, element))
+    {
+        element.function = RPNElement::FUNCTION_UNKNOWN;
+    }
+
+    rpn.emplace_back(std::move(element));
+}
+
+bool ANNCondition::traverseAtomAST(const ASTPtr & node, RPNElement & out)
+{
+    // Match Functions
+    if (const auto * function = node->as<ASTFunction>())
+    {
+        // Set the name
+        out.func_name = function->name;
+
+        if (function->name == "L1Distance" ||
```

Comment:
This should really be a `switch` on function->name.

Reply:
we have several conditions for one action. I'm not sure if `switch` fits here.